src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/InvokerUtil.cs

```diff
@@ -150,8 +166,81 @@ internal static InvokeDelegate GenerateInvokeDelegate(MethodInfo method, out int
 
                     return result;
                 };
+            }
+
+            private static InvokeDelegate GenerateInvokeDelegateInternalWithExpressions(MethodInfo method, out int inputParameterCount, out int outputParameterCount)
+            {
+                inputParameterCount = 0;
+                outputParameterCount = 0;
+                ParameterInfo[] parameters = method.GetParameters();
+                bool returnsValue = method.ReturnType != typeof(void);
+
+                var targetParam = Expression.Parameter(typeof(object), "target");
+                var inputsParam = Expression.Parameter(typeof(object[]), "inputs");
+                var outputsParam = Expression.Parameter(typeof(object[]), "outputs");
+
+                List<ParameterExpression> variables = new();
+                var result = Expression.Variable(typeof(object), "result");
+                variables.Add(result);
+
+                List<ParameterExpression> outputVariables = new();
+                List<ParameterExpression> invocationParameters = new();
+                List<Expression> expressions = new();
+
+                for (int i = 0; i < parameters.Length; i++)
+                {
+                    Type variableType = parameters[i].ParameterType.IsByRef
+                        ? parameters[i].ParameterType.GetElementType()
+                        : parameters[i].ParameterType;
+                    ParameterExpression variable = Expression.Variable(variableType, $"p{i}");
+
+                    if (ServiceReflector.FlowsIn(parameters[i]))
+                    {
+                        expressions.Add(Expression.Assign(variable, Expression.Convert(Expression.ArrayIndex(inputsParam, Expression.Constant(inputParameterCount)), variableType)));
+                        inputParameterCount++;
+                    }
+
+                    if (ServiceReflector.FlowsOut(parameters[i]))
+                    {
+                        outputParameterCount++;
+                        outputVariables.Add(variable);
+                    }
+
+                    variables.Add(variable);
+                    invocationParameters.Add(variable);
+                }
+
+                var castTargetParam = Expression.Convert(targetParam, method.DeclaringType);
+
+                expressions.Add(returnsValue
+                    ? Expression.Block(Expression.Assign(result, Expression.Convert(Expression.Call(castTargetParam, method, invocationParameters), typeof(object))))
+                    : Expression.Block(Expression.Call(castTargetParam, method, invocationParameters),
```

Comment:
Do you need the Expression.Convert? Is this needed in the case where the return value is a value type so it needs to be boxed? If the return type is a value type, then I would expect it to assign to an object variable without any problem.

Reply:
I just retried without the Expression.Convert. I got the following exception (sorry french) System.ArgumentException: 'Impossible d'utiliser une expression de type 'System.Int32' pour une assignation au type 'System.Object''.
The contract was
```c#
    [System.ServiceModel.ServiceContract]
    public interface ICalculatorService
    {
        [System.ServiceModel.OperationContract]
        [System.ServiceModel.FaultContract(typeof(MathFault))]

        int Divide(int a, int b);
    }
```

Reply:
The French is okay, I used to be reasonably fluent, and I understand what it's saying. I think the issue is because it's a value type needing to be boxed. Is the convert needed for a non-value type?

Reply:
I tracked down the source code and found the logic. If both are the same type, it's fine (doesn't matter if reference type of value type). If they types are different and either is a value type, it fails. If they are both reference types, then is uses Type.IsAssignableFrom. So it looks like as we're assigning to an object, you don't need to convert if it's a reference type, and only if it's a value type. [Code is here](https://github.com/dotnet/runtime/blob/a54ccf39b17586e7c2a38b882e80a3cc62d83d7d/src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs#L919).